#!/usr/bin/env python
PACKAGE="photonfocus_camera"

from dynamic_reconfigure.parameter_generator_catkin import *
from driver_base.msg import SensorLevels

gen = ParameterGenerator()

# ----- Image Size Control -----
gen.add("Width", int_t, SensorLevels.RECONFIGURE_STOP, "Width of the image.",640,0,648);
gen.add("Height", int_t, SensorLevels.RECONFIGURE_STOP, "Height of the image.",460,0,488);
gen.add("OffsetX", int_t, SensorLevels.RECONFIGURE_RUNNING, "Where the image window has its top left corner.",0,0,8)
gen.add("OffsetY", int_t, SensorLevels.RECONFIGURE_RUNNING, "Where the image window has its top left corner.",20,0,28)

# ----- Exposure and FrameRate -----
#gen.add("ExposureTimeAbs", double_t, SensorLevels.RECONFIGURE_RUNNING, "Exposure", 20.0, 0.011, 1677)
expo_enum = gen.enum(
    [
        gen.const("Off", int_t, 0, "Off"),
        gen.const("Timed", int_t, 1, "Timed (to use exposure time raw)"),
        gen.const("TriggerWidth", int_t, 2, "TriggerWidth (trigger must be set"),
        gen.const("IOExposureControl", int_t, 4, "IOExposureControl")
    ],"Exposure Mode")

gen.add("ExposureMode", int_t, SensorLevels.RECONFIGURE_RUNNING, "Exposure Mode", 0, edit_method=expo_enum)
gen.add("ExposureTimeRaw", int_t, SensorLevels.RECONFIGURE_RUNNING, "Exposure Time Raw", 7000, 2, 8560)

#gen.add("ConstantFramerate_CFR", bool_t, SensorLevels.RECONFIGURE_RUNNING, "Enable/Disable Constant Frame Rate", True)
#gen.add("Frametime", double_t, SensorLevels.RECONFIGURE_RUNNING, "Frame Time", 100.0, 0.011, 1677)

# ----- LinLog -----
linlog_enum = gen.enum(
    [
        gen.const("noCompression", int_t, 0, "Turn off compression"), #Off instead of noCompression
        gen.const("LowCompression", int_t, 1, "A low compression"),
        gen.const("NormalCompression", int_t, 2, "A normal compression"),
        gen.const("HighCompression", int_t, 3, "A high compression"),
        gen.const("UserDefined", int_t, 4, "A user defined compression")
    ],"LinLog Compression")

gen.add("LinLog_Mode", int_t, SensorLevels.RECONFIGURE_RUNNING, "A size parameter which is edited via an enum", 1, edit_method=linlog_enum)
gen.add("LinLog_Value1", int_t, SensorLevels.RECONFIGURE_RUNNING, "", 0, 0, 200)
gen.add("LinLog_Value2", int_t, SensorLevels.RECONFIGURE_RUNNING, "", 0, 0, 200)
gen.add("LinLog_Time1", int_t, SensorLevels.RECONFIGURE_RUNNING, "", 0, 0, 1000)
gen.add("LinLog_Time2", int_t, SensorLevels.RECONFIGURE_RUNNING, "", 0, 0, 1000)
gen.add("Voltages_BlackLevelOffset", int_t, SensorLevels.RECONFIGURE_RUNNING, "", 104, 0, 255)
gen.add("Trigger_Interleave", bool_t, SensorLevels.RECONFIGURE_RUNNING, "Enable/Disable Trigger Interleave", False)

# ----- Encoding -----
encoding_enum = gen.enum(
    [
        gen.const("MONO8", str_t, "Mono8", "MONOCHROME 8bit"),
    ],"Image Encoding")

gen.add("Encoding", str_t, SensorLevels.RECONFIGURE_STOP, "Encoding of the images", "Mono8", edit_method=encoding_enum)

# end
exit(gen.generate(PACKAGE, "photonfocus_node", "photonfocus"))
